File: TheaterSuggestions/Java/ScreeningSeatingArrangementSuggestions/src/main/java/org/octo/seatingplacessuggestions/seatingplacesuggestions/AuditoriumSeatingArrangement.java
package org.octo.seatingplacessuggestions.seatingplacesuggestions;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

public record AuditoriumSeatingArrangement(Map<String, Row> rows) {

    public Map<String, Row> getRows() {
        return rows;
    }

    public SeatingOptionIsSuggested suggestSeatingOptionFor(int partyRequested, PricingCategory pricingCategory) {
        for (Row row : rows.values()) {
            var seatingOptionSuggested = row.suggestSeatingOption(partyRequested, pricingCategory);

            if (seatingOptionSuggested.matchExpectation()) {
                return seatingOptionSuggested;
            }
        }
        return new SeatingOptionIsNotAvailable(partyRequested, pricingCategory);
    }

    public AuditoriumSeatingArrangement allocate(List<SeatingPlace> seatingPlaces) {
        Map<String, Row> newRows = new HashMap<>(rows);
        seatingPlaces.forEach(seatingPlace -> {
            Row formerRow = newRows.get(seatingPlace.rowName());
            var newRow = formerRow.allocate(seatingPlace);
            newRows.put(newRow.name(), newRow);
        });
        return new AuditoriumSeatingArrangement(newRows);
    }
}

File: TheaterSuggestions/CSharp/ScreeningSeatingArrangementSuggestions/AuditoriumSeatingArrangement.cs
using Value.Shared;

namespace SeatsSuggestions;

public class AuditoriumSeatingArrangement(Dictionary<string, Row> rows) : Value.ValueType<AuditoriumSeatingArrangement>
{
    public IReadOnlyDictionary<string, Row> Rows => rows;

    public SeatingOptionIsSuggested SuggestSeatingOptionFor(int partyRequested, PricingCategory pricingCategory)
    {
        foreach (var row in rows.Values)
        {
            var seatOptionsSuggested = row.SuggestSeatingOption(partyRequested, pricingCategory);

            if (seatOptionsSuggested.MatchExpectation()) return seatOptionsSuggested;
        }

        return new SeatingOptionIsNotAvailable(partyRequested, pricingCategory);
    }

    protected override IEnumerable<object> GetAllAttributesToBeUsedForEquality()
    {
        return [new DictionaryByValue<string, Row>(Rows.ToDictionary())];
    }

    public AuditoriumSeatingArrangement Allocate(List<SeatingPlace> seatingPlacesSuggested)
    {
        var newRows = new Dictionary<string, Row>(Rows.ToDictionary());

        foreach (var seatingPlaceSuggested in seatingPlacesSuggested)
        {
            newRows[seatingPlaceSuggested.RowName] = newRows[seatingPlaceSuggested.RowName].Allocate(seatingPlaceSuggested);
        }

        return new AuditoriumSeatingArrangement(newRows);
    }
}